andi/andi.py

```diff
@@ -1,62 +1,302 @@
-        ]
-        ) -> Dict[str, Optional[Type]]:
+ContainerOrCallableType = Union[
+    Container[Callable],
+    Callable[[Callable], bool]
+]
+
+
+PlanList = List[Tuple[Callable, Dict[str, Callable]]]
+
+
+class Plan(PlanList):
     """
-    Return a dictionary ``{argument_name: type}`` with types which should
-    be provided.
+    The resultant plan of executing the ``plan`` function.
+    It contains a sequence of steps that should be executed in order because
+    they are dependant among them.
+
+    The plan is itself a list of tuples of type
+    ``Tuple[Callable, Dict[str, Callable]]``
+    corresponding to
+    (callable_to_invoke, (param_name -> callable_to_build_the_param)).
+    """
+
+    @property
+    def dependencies(self) -> PlanList:
+        """
+        The plan required to build the dependencies for the
+        ``plan`` input function/class only.
+        Useful when it is known that not all dependencies for input
+        function/class could be resolved. In such a case, it is convenient to
+        execute the plan for the dependencies and then
+        have a custom execution step for the
+        input function/class where those non-resolvable dependencies are
+        provided by other means.
+
+        It is just the last item of the original plan ``self[:-1]``
+        """
+        return self[:-1]
+
+    @property
+    def final_arguments(self) -> Dict[str, Callable]:
+        """
+         The input function/class argument names and its builders for
+         those arguments for which it was possible to resolve the dependencies.
+
+         Equivalent to ``self[-1][1]``
+        """
+        _, params = self[-1]
+        return params
+
+
+def plan(class_or_func: Callable, *,
+         is_injectable: ContainerOrCallableType,
+         externally_provided: Optional[ContainerOrCallableType] = None,
+         strict=False) -> Plan:
+    """ Plan the sequence of instantiation steps required to fulfill the
+    the arguments of the given function or the arguments of its
+    constructor if a class is given instead. In other words, this function
+    makes dependency injection easy. Type annotations are used
+    to determine with instance must be built to satisfy the dependency.
+
+    The plan is a sequence steps.
+    Each step in the plan is a tuple with:
+
+    * A callable with the
+      class/function that must be built/invoked in this step
+    * A dictionary with all the kwargs required for the
+      key build/invocation process. This dictionary has the argument names as keys
+      and classes/functions required to build them as values.
+
+    The best way to understand a plan is to see how a typical building
+    function would use it to build the instances::
+
+        def build(plan):  # Build all the instances from a plan
+            instances = {}
+            for fn_or_cls, args in plan:
+                kwargs = {arg: instances[arg_cls]
+                          for arg, arg_cls in args.items()}
```

Comment:
I wonder if we should provide a helper for that in andi, even if the code is small (see _get_kwargs below) - as this code is present in every build example.

Reply:
On the other hand, having it here makes the example more clear

Reply:
I would be left it as is for didactic purposes, given that we spect the user of the library to build its own building function.

Reply:
I think we can add this helper in a separate PR anyways.